Towards 3.0 blog post draft (#335)
MDAnalysis/MDAnalysis.github.io: _posts/2023-10-09-towards_3.0.md

---
layout: post
title: Towards MDAnalysis 3.0
---

## Overview

> MDAnalysis is beginning to plan for our 3.0 release! Following discussions at our first [UGM][], we (the MDAnalysis [Core Developers][]) have decided on a set priorities for our this exciting milestone. Our primary goals can be divided into several main themes.

Reviewer: And a set *of* priorities please :)

### A trimmed down core library

> As MDAnalysis continues to grow along with the field of molecular simulation, we are aiming to continue to provide a highly functional, easily maintainable package that can be the foundation of a wide variety of downstream tools.

Reviewer 2: ```suggestion
As MDAnalysis continues to grow along with the field of molecular simulations, we are aiming to continue to provide a highly functional, easily maintainable package that can be the foundation of a wide variety of downstream tools.
```
?

> Following the successful introduction of the [MDAKit ecosystem][] (modelled on the SciKit ecosystem, see the [MDAKits whitepaper][]) as part of our [EOSS4][] work, which provides infrastructure for creating FAIR packages that address specific scientific questions in the analysis of molecular simulation, we are aiming to move some more specialized analyses currently present in the MDAnalysis core library to MDAKits. Several analysis modules, including  hole2, encore, path-similarity, helanal and waterdynamics are being assessed to undergo this transition. Indeed some of these packages are already deployed as MDAKits, and are deprecated in the core library.

Reviewer 2: ```suggestion
Following the successful introduction of the [MDAKit ecosystem][] (modelled on the SciKit ecosystem, see the [MDAKits whitepaper][]) as part of our [EOSS4][] work, which provides infrastructure for creating FAIR packages that address specific scientific questions in the analysis of molecular simulation, we are aiming to move some more specialized analyses currently present in the MDAnalysis core library to MDAKits. Several analysis modules, including  hole2, encore, path-similarity, helanal and waterdynamics are being assessed to undergo this transition. Indeed some of these packages are already deployed as MDAKits, and are deprecated in the core library.
```
Would it be useful to link to FAIR principles for the uninitiated?

This change will allow for a reduction in the core dependencies of MDAnalysis and reduce the maintenance burden of specialist analyses in the core library. This will also empower those developing specialist analysis code to rapidly develop and prototype their code in well supported downstream packages. Additionally, core developers will be able to spend more cycles on core competencies, fixing critical issues and planning future improvements, ensuring that MDAnalysis can be easily and effectively maintained for years to come.

### Enhancements

We have identified several areas that we would like to target for future improvement of MDAnalysis across three key themes.

* Improving interoperability
    - ....

* Support for streaming and cloud storage
    - ....

* Enhanced parallelism and performance

>     - ...

Reviewer: Can I just ask which points this is meant to cover -- is it the current parallel analysis project?

Author: Yes, + distopia, I will make it explicit.

> 

Reviewer: It would be great to tie EOSS 6 to this roadmap yes -- could we please add guessers/chemical contexts in here? While they didn't come up in the UGM planning meeting, we did get feedback during the 3.0 panel asking for something similar. And if we keep encouraging MDAKits and more diverse analyses, they'll become more and more useful IMO.

> 

Reviewer: I think I initially grouped them as "scientifically aware operations" and lumped in guessers, cheminformatics, and units, but since we're not properly doing units that may not be the best name.

> 

Author: @lilyminium I have tried to link with EOSS6 proposal now. Have a look and see if you like it.

We aim to lay the groundwork for these future directions in our work for MDAnalysis 3.0.

### User facing API changes 

We are also planning to clear up some API inconsistencies scattered across the library that require an API break, including some instances where coordinate readers behave differently relative to each other, renaming some arguments to clearer names and addressing other miscellanea. Additionally, we are aiming to provide a more consistent and usable low-level Cython API surface that compiled extensions can hook into. We are also investigating the fesability of improving consistency of precision handling in various parts of the MDAnalysis core library, as the use of double and mixed precision has been identified as a key performance bottleneck.

### Improving org structure and formalising roles and responsibilities. 

> We have identified that assigning direct roles and responsibilities across our coredevs is a change that will be beneficial for the health of our community going forward.

Reviewer 2: ```suggestion
We have identified that assigning direct roles and responsibilities across our [Core Developers][] is a change that will be beneficial for the health of our community going forward.
```
I prefer the full spelling in a blog post with wide audience.

> To this end we are workshopping a more formal org structure that will make our internal processes more transparent and allow more functional redundancy in key roles. Watch this space!

Reviewer: Thanks for including this, I guess it's not part of the code and as part of 3.0 but IMO it's definitely something we identified as an organisational priority for MDA as a whole!

> To this end we are workshopping a more formal org structure that will make our internal processes more transparent and allow more functional redundancy in key roles. Watch this space!

Reviewer 2: ```suggestion
To this end we are workshopping a more formal organization structure that will make our internal processes more transparent and allow more functional redundancy in key roles. Watch this space!
```

---
layout: post
title: Towards MDAnalysis 3.0
---

## Overview

MDAnalysis is beginning to plan for our 3.0 release! Following discussions at our first [UGM][], we (the MDAnalysis [Core Developers][]) have decided on a set priorities for this exciting milestone (detailed below). Our primary goals can be divided into several main themes.

### A trimmed down core library

As MDAnalysis continues to grow along with the field of molecular simulations, we are aiming to continue to provide a highly functional, easily maintainable package that can be the foundation of a wide variety of downstream tools.

> Following the successful introduction of the [MDAKits ecosystem][] (modelled on the SciKit ecosystem, see the [MDAKits whitepaper][]) as part of our [EOSS4][] work, which provides infrastructure for creating [FAIR][] packages that address specific scientific questions in the analysis of molecular simulation, we are aiming to move some more specialized analyses currently present in the MDAnalysis core library to MDAKits. Several analysis modules, including  hole2, encore, path-similarity, helanal and waterdynamics are being assessed to undergo this transition. Indeed some of these packages are already deployed as MDAKits, and are deprecated in the core library.

Reviewer 2: ```suggestion
Following the successful introduction of the [MDAKits ecosystem][] (modelled on the SciKit ecosystem, see the [MDAKits whitepaper][]) as part of our [EOSS4][] work, which provides infrastructure for creating [FAIR][] packages that address specific scientific questions in the analysis of molecular simulation, we are aiming to move some more specialized analyses currently present in the MDAnalysis core library to [MDAKits][]. Several analysis modules, including  hole2, encore, path-similarity, helanal and waterdynamics are being assessed to undergo this transition. Indeed some of these packages are already deployed as [MDAKits][], and are deprecated in the core library.
```

This change will allow for a reduction in the core dependencies of MDAnalysis and reduce the maintenance burden of specialist analyses in the core library. This will also empower those developing specialist analysis code to rapidly develop and prototype their code in well supported downstream packages. Additionally, core developers will be able to spend more cycles on core competencies, fixing critical issues and planning future improvements, ensuring that MDAnalysis can be easily and effectively maintained for years to come.

### Enhancements

We have identified several areas that we would like to target for future improvement of MDAnalysis across four key themes.

* Improving interoperability
    - Add support for new formats
    - Improve interoperability with chemoinformatics and simulation software

* Scientifically aware operations

>     - Improve guessers to enable context dependent chemical information to be easily inferred

Reviewer: ```suggestion
    - Improve guessers to enable context-dependent chemical information to be easily inferred
```

>     - Make MDAnalysis more chemoinformatics aware

Reviewer: ```suggestion
    - Make MDAnalysis more chemoinformatics-aware
```

>     - Improve support for units and unit systems

Reviewer: I think there was a discussion about this and how unit-ed arrays are a chore to work with, and that the best solution would be better communication of what units we use i.e. documentation (cc @richardjgowers)

Author: This is non-committal enough to give us flexibility in how we approach I think.

* Support for streaming and cloud storage
    - Add support for generic streams in file readers

>     - Enable reading of trajectories and topologies directly from cloud storage ()

Reviewer: Extra `()`?

* Enhanced parallelism and performance
    - Improve performance of key distance routines with [distopia][]
    - Create a compiled backend for ASCII topology and coordinate readers

>     - Redesign precision model in MDAnalysis internals.

Reviewer: ```suggestion
    - Redesign precision model in MDAnalysis internals to guarantee user-controlled precision
```

(iirc) just an expansion / explanation why we're doing this

### Improving organization structure and formalising roles and responsibilities. 

We have identified that assigning direct roles and responsibilities across our [Core Developers][] is a change that will be beneficial for the health of our community going forward.
To this end we are workshopping a more formal organization structure that will make our internal processes more transparent and allow more functional redundancy in key roles. Watch this space!

If you have any feedback on this roadmap or would like to discuss our future plans with us, contact us on our [Discord][] or [mailing list][].

– Hugo MacDermott-Opeskin on behalf of the MDAnalysis [Core Developers][]

[Core Developers]: {% link about.md %}#mdanalysis-core-developers
[Discord]: https://discord.gg/juV3QbAMru
[distopia]: https://github.com/MDAnalysis/distopia
[EOSS4]: {{ site.baseurl }}{% post_url 2021-08-31-CZI-EOSS4 %}
[FAIR]: https://www.go-fair.org/fair-principles/
[GSOC project]: https://www.mdanalysis.org/blog/page4/#egor-marin-implementation-of-parallel-analysis-in-mdanalysis
[mailing list]: https://groups.google.com/g/mdnalysis-discussion

> [MDAKits ecosystem]: https://github.com/MDAnalysis/mdakits

Reviewer 2: ```suggestion
[MDAKits ecosystem]: https://github.com/MDAnalysis/mdakits
[MDAKits]: https://github.com/MDAnalysis/mdakits
```

---
layout: post
title: Towards MDAnalysis 3.0
---

## Overview

MDAnalysis is beginning to plan for our 3.0 release! Following discussions at our first [UGM][], we (the MDAnalysis [Core Developers][]) have decided on a set priorities for this exciting milestone (detailed below). Our primary goals can be divided into several main themes.

### A trimmed down core library

As MDAnalysis continues to grow along with the field of molecular simulations, we are aiming to continue to provide a highly functional, easily maintainable package that can be the foundation of a wide variety of downstream tools.

> Following the successful introduction of the [MDAKits ecosystem][] (modelled on the SciKit ecosystem, see the [MDAKits whitepaper][]) as part of our [EOSS4][] work, which provides infrastructure for creating [FAIR][] packages that address specific scientific questions in the analysis of molecular simulation, we are aiming to move some more specialized analyses currently present in the MDAnalysis core library to [MDAKits][]. Several analysis modules, including  hole2, encore, path-similarity, helanal and waterdynamics are being assessed to undergo this transition. Indeed some of these packages are already deployed as [MDAKits][], and are deprecated in the core library.

Reviewer 2: ```suggestion
Following the successful introduction of the [MDAKits ecosystem][] (modelled on the SciKit ecosystem, see the [MDAKits whitepaper][]) as part of our [EOSS4][] work, which provides infrastructure for creating [FAIR][] packages that address specific scientific questions in the analysis of molecular simulation, we are aiming to move some more specialized analyses currently present in the MDAnalysis core library to [MDAKits][]. Several analysis modules, including  `hole2`, `encore`, `path-similarity`, `helanal` and `waterdynamics` are being assessed to undergo this transition. Indeed some of these packages are already deployed as [MDAKits][], and are deprecated in the core library.
```

> Following the successful introduction of the [MDAKits ecosystem][] (modelled on the SciKit ecosystem, see the [MDAKits whitepaper][]) as part of our [EOSS4][] work, which provides infrastructure for creating [FAIR][] packages that address specific scientific questions in the analysis of molecular simulation, we are aiming to move some more specialized analyses currently present in the MDAnalysis core library to [MDAKits][]. Several analysis modules, including  hole2, encore, path-similarity, helanal and waterdynamics are being assessed to undergo this transition. Indeed some of these packages are already deployed as [MDAKits][], and are deprecated in the core library.

Reviewer 3: We should point to our scipy paper (doi https://doi.org/10.25080/gerudo-f2bc6f59-00a) rather than the whitepaper

Author: done.

This change will allow for a reduction in the core dependencies of MDAnalysis and reduce the maintenance burden of specialist analyses in the core library. This will also empower those developing specialist analysis code to rapidly develop and prototype their code in well supported downstream packages. Additionally, core developers will be able to spend more cycles on core competencies, fixing critical issues and planning future improvements, ensuring that MDAnalysis can be easily and effectively maintained for years to come.

> 

Reviewer 2: Good point. I think, as you say, that moved stuff needs to be maintained, at least for a while. Otherwise MDAKits will be seen mainly as a way to dump code from core and leave it there. I think we have to do the initial maintenance cycle, but totally agree with what you wrote. If someone is willing to take ownership, that great. 

However, what happens if the new owner let it die? That's maybe something to discuss.

> 

Reviewer 4: See discussion on business. Let's wait until Monday if there are other opinions, otherwise we should go ahead with the above and the support of the coredevs.

> 

Reviewer 4: @MDAnalysis/coredevs support the changes above ✅

### Enhancements

We have identified several areas that we would like to target for future improvement of MDAnalysis across four key themes.

> * Improving interoperability

Reviewer 3: We could specifically mention converters as MDAKits as a work-in-progress here, since I know this is something @IAlibay has been/is planning to work on

Author: ah yes will do.
